go/chat/search/storage.go

```diff
@@ -68,18 +69,268 @@ func (a *aliasEntry) remove(token string) bool {
+	return &tokenBatch{
+		convID: convID,
+		tokens: make(map[string]*tokenEntry),
+	}
+}
+
+type mdBatch struct {
+	convID chat1.ConversationID
+	md     *indexMetadata
+}
+
+type batchingStore struct {
+	utils.DebugLabeler
+	sync.Mutex
+
+	uid        gregor1.UID
+	mdb        *libkb.JSONLocalDb
+	edb        *encrypteddb.EncryptedDB
+	keyFn      func(ctx context.Context) ([32]byte, error)
+	aliasBatch map[string]*aliasEntry
+	tokenBatch map[string]*tokenBatch
+	mdBatch    map[string]*mdBatch
+	flushMu    sync.Mutex
+}
+
+func newBatchingStore(log logger.Logger, uid gregor1.UID,
+	keyFn func(ctx context.Context) ([32]byte, error), edb *encrypteddb.EncryptedDB,
+	mdb *libkb.JSONLocalDb) *batchingStore {
+	b := &batchingStore{
+		DebugLabeler: utils.NewDebugLabeler(log, "Search.batchingStore", false),
+		uid:          uid,
+		keyFn:        keyFn,
+		edb:          edb,
+		mdb:          mdb,
+	}
+	b.Lock()
+	b.resetLocked()
+	b.Unlock()
+	return b
+}
+
+func (b *batchingStore) resetLocked() {
+	b.aliasBatch = make(map[string]*aliasEntry)
+	b.tokenBatch = make(map[string]*tokenBatch)
+	b.mdBatch = make(map[string]*mdBatch)
+}
+
+func (b *batchingStore) GetTokenEntry(ctx context.Context, convID chat1.ConversationID,
+	token string) (res *tokenEntry, err error) {
+	b.Lock()
+	defer b.Unlock()
+	batch, ok := b.tokenBatch[convID.String()]
+	if ok {
+		return batch.tokens[token], nil
+	}
+	key, err := tokenKey(ctx, b.uid, convID, token, b.keyFn)
+	if err != nil {
+		return nil, err
+	}
+	res = new(tokenEntry)
+	found, err := b.edb.Get(ctx, key, res)
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, nil
+	}
+	return res, nil
+}
+
+func (b *batchingStore) PutTokenEntry(ctx context.Context, convID chat1.ConversationID,
+	token string, te *tokenEntry) (err error) {
+	b.Lock()
+	defer b.Unlock()
+	key := convID.String()
+	batch, ok := b.tokenBatch[key]
+	if !ok {
+		batch = newTokenBatch(convID)
+	}
+	batch.tokens[token] = te
+	b.tokenBatch[key] = batch
+	return nil
+}
+
+func (b *batchingStore) RemoveTokenEntry(ctx context.Context, convID chat1.ConversationID,
+	token string) {
+	b.Lock()
+	defer b.Unlock()
+	batch, ok := b.tokenBatch[convID.String()]
+	if ok {
+		delete(batch.tokens, token)
+	}
+	key, err := tokenKey(ctx, b.uid, convID, token, b.keyFn)
+	if err != nil {
+		b.Debug(ctx, "RemoveTokenEntry: failed to get tokenkey: %s", err)
+		return
+	}
+	if err := b.mdb.Delete(key); err != nil {
+		b.Debug(ctx, "RemoveTokenEntry: failed to delete key: %s", err)
+	}
+}
+
+func (b *batchingStore) GetAliasEntry(ctx context.Context, alias string) (res *aliasEntry, err error) {
+	b.Lock()
+	defer b.Unlock()
+	var ok bool
+	if res, ok = b.aliasBatch[alias]; ok {
+		return res, nil
+	}
+	key, err := aliasKey(ctx, alias, b.keyFn)
+	if err != nil {
+		return res, err
+	}
+	res = new(aliasEntry)
+	found, err := b.edb.Get(ctx, key, res)
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, nil
+	}
+	return res, nil
```

Comment:
populate aliasBatch

Reply:
It isn't a cache, it is supposed to mirror what you would have gotten off the disk.